src/mgr/DaemonState.h

```diff
@@ -261,9 +261,13 @@ class DaemonStateIndex
   template<typename Callback, typename...Args>
   auto with_daemons_by_server(Callback&& cb, Args&&... args) const ->
     decltype(cb(by_server, std::forward<Args>(args)...)) {
-    std::shared_lock l{lock};
-    
-    return std::forward<Callback>(cb)(by_server, std::forward<Args>(args)...);
+    decltype(by_server) by_server_copy;
+    {
+        // Don't hold the lock any longer than necessary
+        std::shared_lock l{lock};
+        by_server_copy = by_server;
```

Comment:
I agree with the idea behind the patch.
What worries is this little detail. Before fixing the `cb` was supplied with the `by_server` being the member of the class; after – just its local, on-stack copy. This 1) changes the life-time of the callback's first argument and 2) makes any modification local. Unfortunately, as the method is a template, it cannot be determined there is no user taking this parameter by reference, particularly non-`const` reference.